Data-sharing permission - collect, validate, persist, display

labkey-adam opened the pull request:
#### Rationale
FDA MyStudies provides the option to ask participants if they consent to their data being shared with 3rd parties for research. This change to the Response server collects, validates, persists, and makes this flag available in grids and downstream analysis. https://docs.google.com/document/d/15_G_Po8e4tMuDyxKE5r1Ocw7t1Hmdfw95KOcgd3Jucc/edit#

#### Changes
* Collect, validate, and persist a new, required "allowDataSharing" parameter at enrollment time (enroll.api). Valid values are "true", 'false", "NA".
* Make AllowDataSharing column available in response grids, export, APIs, queries, etc.
* Fix existing automated tests and add an explicit test of "allowDataSharing" valid & invalid values.
* Graft "Token" column into Participant table for easy joining of the enrollment token into response lists.
* Expose enum tables to the user schema so they show up in schema browser, etc.

labkey-adam committed:
Data-sharing permission - collect, validate, persist, display

labkey-adam committed:
Fix existing tests - add "allowDataSharing" parameter to EnrollParticipantCommand and pass in valid values

labkey-adam committed:
Add "allowDataSharing" validation test

labkey-adam committed:
Add valid "allowDataSharing" parameter to assignToken() test helper

labkey-adam committed:
Encode spaces

Sigmonia reviewed (approved):
Question that I don't see a home for:

Since the enrollment is essentially a write once action, can the user ever revoke/alter their consent? at least going forward from the change.

labkey-adam commented:
Good question. I'll run this past BTC and HPHC, but there's been no discussion about revoking data sharing consent. There is the big hammer of "withdraw from study and delete all my data."

labkey-adam commented:
I confirmed with BTC that there's no way for a participant to revoke this consent. BTC's opinion is that we'd need to keep a dated history, since some of the data would be consented for sharing and some wouldn't. One could argue whether that's strictly necessary, but I wouldn't pursue any of this unless we get strong requirements from investigators.

labkey-adam committed:
Join enrollment token into Participant table to make it available to data anaylysts in the reponse lists

Fix enum tables - show all enum tables in schema browser, FK on ParticipantPropertyType, rename ParticipantPropertyType table for consistency

labkey-adam committed:
Retrieve allowDataSharing flag after enrollment

labkey-adam committed:
Fix schema name. Clear out previous exception message.

labkey-adam committed:
Set tables to read-only. Tweak the test.

labkey-adam committed:
Cleaner approach for grafting Token into ParticipantTable

labkey-adam committed:
Add logging to help track down missing column

labkey-adam committed:
Bump column count down -- _labkeyurl_ParticipantId is no longer showing up
